src/cache.rs

```diff
@@ -340,23 +339,26 @@ impl Cache {
 
     /// Typecheck an entry of the cache and update its state accordingly, or do nothing if the
     /// entry has already been typechecked. Require that the corresponding source has been parsed.
+    /// If the source contains imports, recursively typecheck on the imports too.
     pub fn typecheck(
         &mut self,
         file_id: FileId,
         global_env: &eval::Environment,
     ) -> Result<CacheOp<()>, CacheError<TypecheckError>> {
-        if !self.terms.contains_key(&file_id) {
-            return Err(CacheError::NotParsed);
-        }
-
-        // After self.parse(), the cache must be populated
-        let (t, state) = self.terms.get(&file_id).unwrap();
+        let (t, state) = self.terms.get(&file_id).ok_or(CacheError::NotParsed)?;
 
         if *state > EntryState::Typechecked {
             Ok(CacheOp::Cached(()))
         } else if *state >= EntryState::Parsed {
             type_check(t, global_env, self)?;
             self.update_state(file_id, EntryState::Typechecked);
+
+            if let Some(imports) = self.imports.get(&file_id).cloned() {
+                for f in imports.into_iter() {
+                    self.typecheck(f, global_env)?;
+                }
+            }
```

Comment:
Is it not better to typecheck deeper imports before `import` callers? I mean, put the for loop before `type_check(t, global_env, self)?;`. At least, I think we update the state to early: what append if the typechecking of a deep import fail?

Reply:
I think I had to do this to make operations on cyclic imports to succeed. Say we are typechecking `a`, and there is a cyclic import `a imports b, b imports a`. And say do as before: start typechecking `a`, then typecheck `b`, and only then finally set the state of `a` to `Typechecked`. Then, in the for loop, we will typecheck `b`. `b` will do the same thing, typecheck the content of `b`, and then typecheck imports. Including `a`. Because `a` is still in the state `Parsed`, it will go again, which creates the exact kind of loop we had in `type_check`, but just now it is in the cache layer.

Here, the `EntryState` is to understand as "is a specific operation has been performed on this specific entry". `Typechecked` would mean "this particular file has been typechecked", which is indeed what typechecks do, without telling anything about imports. Under this interpretation, we are always in a correct state. But indeed it must be documented clearly so that users of the cache don't make false assumption.

One alternative is to add intermediate states like `Typechecking`, `Transforming`, and so on. I don't know if this is very useful, but it's not very expensive to do though.

Reply:
I expanded the documentation on `EntryState`. I think it should do the trick for now, if we ever need to differentiate between "the entry only is typechecked" and "the entry and all of its transitive dependencies are typechecked" at some point, we can add the missing states then. What do yo think @francois-caddet ?

Reply:
The issue is:
When the `EntryState` is `Resolved`, the function just ignore the import and do not resolve it. Right now it's not realy an issue because an error in any file is fatal. But if later we would like to add a recover on error system, it will not work as expected. It can be managed by checking `Cache::imports` tree and the coresponding state. But we have to document it, because if not done, we will have very wear evaluation errors.

Reply:
Good point, it may be more natural for `Typechecked` to mean everything went fine, imports included.